src/components/fx/click.js

```diff
@@ -14,12 +14,30 @@ module.exports = function click(gd, evt, subplot) {
         hover(gd, evt, subplot, true);
     }
 
-    function emitClick() { gd.emit('plotly_click', {points: gd._hoverdata, event: evt}); }
+    function emitClick(data) { gd.emit('plotly_click', {points: data, event: evt}); }
 
-    if(gd._hoverdata && evt && evt.target) {
-        if(annotationsDone && annotationsDone.then) {
-            annotationsDone.then(emitClick);
-        } else emitClick();
+    var clickmode = gd._fullLayout.clickmode;
+    var data;
+    if(evt && evt.target) {
+        if(gd._hoverdata) {
+            data = gd._hoverdata;
+        } else if(clickmode.indexOf('anywhere') > -1) {
+            if(gd._fullLayout.geo) {
```

Comment:
This logic means that if the plot has a geo subplot, we're assuming the click was on that, even if the plot has other subplots too. But we can do a lot better than that, and there are more subplot types than just geo and 2D cartesian. And fortunately we already have a mechanism to detect this: the third arg to `click` is `subplot` and it should tell us exactly which subplot the click came from.

2D cartesian, ternary, and polar subplots report this correctly (eg `'x2y3'`, `'ternary2'`, `'polar3'`), geo doesn't but it should be easy to fix that. Then what we need to do is find the actual subplot object, and depending on its type calculate the appropriate coordinates within that particular subplot.

Pie, sankey, and funnelarea also all reach this point but don't give a subplot. We could have them give the trace number I guess, but is there anything useful to report for them? Just the raw coordinates within the plot?

3D, parcoords, and parcats don't even get here, I'm happy to ignore those for now.

Reply:
Is that something that you intend to do in a second round or within this PR. If it is something I should fix, then I'd like a short example ideally for each plot type as I am not familiar with most of the ones you listed.

Reply:
Please see https://rreusser.github.io/plotly-mock-viewer/#plot_types as well as https://rreusser.github.io/plotly-mock-viewer/#polar_subplots which render https://github.com/plotly/plotly.js/blob/master/test/image/mocks/plot_types.json and
https://github.com/plotly/plotly.js/blob/master/test/image/mocks/polar_subplots.json.